presto-common/src/main/java/com/facebook/presto/common/zorder/ZOrderIntegers.java

```diff
@@ -0,0 +1,201 @@
+        }
+
+        if (input.size() > encodingBits.size()) {
+            throw new IllegalArgumentException(format("Input list size (%d) does not match encoding bits list size (%d).",
+                    input.size(), encodingBits.size()));
+        }
+
+        for (int index = 0; index < input.size(); index++) {
+            if (input.get(index) < 0) {
+                throw new IllegalArgumentException("Our current Z-Ordering implementation does not support negative input numbers.");
+            }
+        }
+    }
+
+    private void checkInputValidity(List<Integer> input, Type type)
+            throws IllegalArgumentException
+    {
+        checkInputValidity(input);
+
+        int maximumBits = 0;
+        switch (type) {
+            case INT:
+                maximumBits = 32;
+                break;
+            case LONG:
+                maximumBits = 64;
+                break;
+            default:
+                break;
+        }
+
+        if (totalBitLength > maximumBits) {
+            throw new IllegalArgumentException(format("The z-address type specified is not large enough to hold %d bits.", totalBitLength));
+        }
+    }
+
+    @Override
+    public byte[] encodeToByteArray(List input)
+            throws IllegalArgumentException
+    {
+        checkInputValidity(input);
+
+        byte[] address = new byte[(totalBitLength + 7) >> 3];
+        int bitIndex = totalBitLength - 1;
+        // Interweave input bits into address from the most significant bit to preserve data locality
+        for (int bitsProcessed = 1; bitsProcessed <= maxBitLength; bitsProcessed++) {
+            for (int index = 0; index < input.size(); index++) {
+                if (bitsProcessed > encodingBits.get(index)) {
+                    continue;
+                }
+                int bitPosition = encodingBits.get(index) - bitsProcessed;
+                byte maskedBit = (byte) (((int) input.get(index) >> bitPosition) & 1);
+                address[bitIndex >> 3] |= maskedBit << (bitIndex & 7);
+                bitIndex--;
+            }
+        }
+        return address;
+    }
+
+    @Override
+    public long encodeToNumber(List input, Type type)
+    {
+        checkInputValidity(input, type);
+
+        long address = 0;
+        int bitIndex = totalBitLength - 1;
+        // Interweave input bits into address from the most significant bit to preserve data locality
+        for (int bitPosition = 1; bitPosition <= maxBitLength; bitPosition++) {
+            for (int index = 0; index < input.size(); index++) {
+                if (bitPosition > encodingBits.get(index)) {
+                    continue;
+                }
+                int inputBitPos = encodingBits.get(index) - bitPosition;
+                long maskedBit = (((long) input.get(index) >> inputBitPos) & 1);
+                address |= maskedBit << bitIndex;
+                bitIndex--;
+            }
+        }
+        return address;
+    }
```

Comment:
Please try to use the result from function encodeToByteArray, instead of writing the same logic twice.